Civi/Api4/Action/PaymentProcessor/PaymentProcessorSaveTrait.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/*
+ +--------------------------------------------------------------------+
+ | CiviCRM version 5                                                  |
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC (c) 2004-2019                                |
+ +--------------------------------------------------------------------+
+ | This file is a part of CiviCRM.                                    |
+ |                                                                    |
+ | CiviCRM is free software; you can copy, modify, and distribute it  |
+ | under the terms of the GNU Affero General Public License           |
+ | Version 3, 19 November 2007 and the CiviCRM Licensing Exception.   |
+ |                                                                    |
+ | CiviCRM is distributed in the hope that it will be useful, but     |
+ | WITHOUT ANY WARRANTY; without even the implied warranty of         |
+ | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
+ | See the GNU Affero General Public License for more details.        |
+ |                                                                    |
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC (c) 2004-2019
+ * $Id$
+ *
+ */
+
+namespace Civi\Api4\Action\PaymentProcessor;
+
+/**
+ * @inheritDoc
+ */
+trait PaymentProcessorSaveTrait {
+
+  /**
+   * @inheritDoc
+   */
+  protected function writeObjects($items) {
+    foreach ($items as &$item) {
+      if (empty($item['domain_id'])) {
+        $item['domain_id'] = \CRM_Core_Config::domainID();
+      }
+      if (empty($item['financial_account_id'])) {
+        $item['financial_account_id'] = \CRM_Financial_BAO_PaymentProcessor::getDefaultFinancialAccountID();
+      }
+      if (empty($item['payment_instrument_id'])) {
+        $paymentProcessorTypes = \Civi\Api4\PaymentProcessorType::get()->addWhere('id', '=', $item['payment_processor_type_id'])->execute();
+        foreach ($paymentProcessorTypes as $paymentProcessorType) {
+          $item['payment_instrument_id'] = $paymentProcessorType['payment_instrument_id'];
+        }
+      }
+    }
```

Comment:
I think setting these field defaults would be better addressed by a SpecProvider, and we should avoid overriding the `writeObjects` function if at all possible.

See for example
https://github.com/civicrm/civicrm-core/blob/5.19/Civi/Api4/Service/Spec/Provider/NavigationSpecProvider.php#L50-L50

Reply:
@colemanw how would you handle the situation say in L58 where setting the payment instrument is dependant on if they have passed in one or if not then get it from the Payment Processor Type ?

Reply:
Good question! I don't think that can be done with SpecProvider, you make a good point there.

However, as it's written this function is problematic because it assumes a complete record is being passed in but (in the case of update) that may not be the case.

My vote is to move this logic to the BAO (and only fire it if `payment_processor_id` is passed in) :)

Reply:
Have rebased this now that the code has shifted to the BAO level